### ISSUE TITLE
System.InvalidOperationException : Collection was modified; enumeration operation may not execute.

### ISSUE THREAD

**Issue by BottlecapDave** [anshooarora/extentreports-csharp#103]
Not sure if to create a new issue or attach a comment to https://github.com/anshooarora/extentreports-csharp/issues/75. When using ExtentReports with xunit and tests executed in parallel, I'm sometimes receiving the following stack trace and exception:

Result StackTrace:	
at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.Model.Test.UpdateTestStatusRecursive(Test test)
   at AventStack.ExtentReports.Model.Test.End()
   at AventStack.ExtentReports.Model.Test.<>c.<EndChildTestsRecursive>b__73_0(Test x)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.Model.Test.EndChildTestsRecursive(Test test)
   at AventStack.ExtentReports.Model.Test.End()
   at AventStack.ExtentReports.Model.Report.EndTest(Test test)
   at AventStack.ExtentReports.Model.Report.<CollectRunInfo>b__19_0(Test test)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.Model.Report.CollectRunInfo()
   at AventStack.ExtentReports.Model.Report.AddLog(Test test, Log log)
   at AventStack.ExtentReports.ExtentTest.AddLog(Log evt)
   at AventStack.ExtentReports.ExtentTest.Log(Status status, String details, MediaEntityModelProvider provider)
   at AventStack.ExtentReports.ExtentTest.Pass(String details, MediaEntityModelProvider provider)

I'm not sure if this is any help, but I'm using ExtentReports with a gherkin structure as I'm using it with Specflow and xUnit. The creation of the step node and the passing/failing of said node is done within the AfterStep callback.

**Comment by anshooarora** [comment]
Version?

**Comment by BottlecapDave** [comment]
I am currently using 3.0.2. I see that there is a newer version out, so will try that

**Comment by anshooarora** [comment]
Yes, version 3.1 launched yesterday. Please check.

**Comment by anshooarora** [comment]
Are you able to reproduce this?  If not, can this be closed?

**Comment by anshooarora** [comment]
Closing this for now. Please reopen or create a new issue if you encounter this again.

**Comment by andrewlaser** [comment]
Currently the latest C# ExtentReports 3.1.3.
Got the same random issue when running tests in parallel.
```
StackTrace:    at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.SessionStatusStats.ExtractStandardTestCountsTestStrategy(Test test)
   at AventStack.ExtentReports.SessionStatusStats.<ExtractStandardTestCountsTestStrategy>b__108_1(Test x)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.SessionStatusStats.ExtractStandardTestCountsTestStrategy(Test test)
   at AventStack.ExtentReports.SessionStatusStats.ExtractCounts(Test test)
   at AventStack.ExtentReports.SessionStatusStats.<UpdateCounts>b__102_0(Test x)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.SessionStatusStats.UpdateCounts()
   at AventStack.ExtentReports.SessionStatusStats.Refresh(List`1 testCollection)
   at AventStack.ExtentReports.Model.Report.CollectRunInfo()
   at AventStack.ExtentReports.Model.Report.AddLog(Test test, Log log)
   at AventStack.ExtentReports.ExtentTest.AddLog(Log evt)
   at AventStack.ExtentReports.ExtentTest.Log(Status status, String details, MediaEntityModelProvider provider)
   at AventStack.ExtentReports.ExtentTest.Debug(String details, MediaEntityModelProvider provider)
   at RS.Framework.Pages.Page.FluentlyWaitForElementToBePresent(By element, Int32 timeoutSeconds) in C:\BuildAgent\work\c9058bdd19501473\RS.Framework\Pages\Page.cs:line 303
   at RS.Framework.Pages.Admin.AdminHeader..ctor(RsWebDriver rsWebDriver) in C:\BuildAgent\work\c9058bdd19501473\RS.Framework\Pages\Admin\AdminHeader.cs:line 53
   at RS.Framework.Pages.Admin.Test.ToBeProcessedPage..ctor(RsWebDriver rsWebDriver) in C:\BuildAgent\work\c9058bdd19501473\RS.Framework\Pages\Admin\Test\ToBeProcessedPage.cs:line 20
   at RS.Tests.Tests.CSC_CustomerService_Collections.Ops.NoSiACommunicationCases.NoSiATemplateLetter() in C:\BuildAgent\work\c9058bdd19501473\RS.Tests\Tests\CSC_CustomerService_Collections\Ops\NoSiACommunicationCases.cs:line 167 
```

Please, provide at least some temporary workaround.
I'm developing a proof of concept for my current project.
Won't be able to persuade my management to buy a license for something that doesn't work :(

**Comment by MattKeenum** [comment]
I am running v3.1.3 and I ran into this issue when running in parallel. I am attempting to use locks to prevent this from happening, but I would prefer not too. 

```
Result Message:	System.InvalidOperationException : Collection was modified; enumeration operation may not execute.

Result StackTrace:	
at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.Model.Test.UpdateTestStatusRecursive(Test test)
   at AventStack.ExtentReports.Model.Test.End()
   at AventStack.ExtentReports.Model.Report.EndTest(Test test)
   at AventStack.ExtentReports.Model.Report.<CollectRunInfo>b__19_0(Test test)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at AventStack.ExtentReports.Model.Report.CollectRunInfo()
   at AventStack.ExtentReports.Model.Report.AddLog(Test test, Log log)
   at AventStack.ExtentReports.ExtentTest.AddLog(Log evt)
   at AventStack.ExtentReports.ExtentTest.Log(Status status, String details, MediaEntityModelProvider provider)
   at AventStack.ExtentReports.ExtentTest.Pass(String details, MediaEntityModelProvider provider)
```
